skyportal/models/user_notification.py

```diff
@@ -90,30 +114,172 @@ def send_slack_notification(mapper, connection, target):
+        )
+    except Exception as e:
+        log(f"Error sending slack notification: {e}")
+
+
+@event.listens_for(UserNotification, 'after_insert')
+def send_email_notification(mapper, connection, target):
+
+    if not email:
+        return
+
+    if not target.user:
+        return
+
+    if not target.user.contact_email:
+        return
+
+    if not target.user.preferences:
+        return
+
+    prefs = target.user.preferences.get('notifications')
+
+    if not prefs:
+        return
+
+    ressource_type = None
+    if "favorite_sources" not in target.notification_type:
+        ressource_type = target.notification_type
+    elif "favorite_sources" in target.notification_type:
+        ressource_type = "favorite_sources"
+
+    if not prefs.get(ressource_type, False):
+        return
+    if not prefs[ressource_type].get("email", False):
+        return
+    if not prefs[ressource_type]['email'].get("active", False):
+        return
+
+    subject = None
+    body = None
+
+    if ressource_type == "sources":
+        subject = f"{cfg['app.title']} - New followed classification on a source"
+        body = f'{target.text} ({get_app_base_url()}{target.url})'
+
+    elif ressource_type == "gcn_events":
+        subject = f"{cfg['app.title']} - New GCN Event with followed notice type"
+        body = f'{target.text} ({get_app_base_url()}{target.url})'
+
+    elif ressource_type == "facility_transactions":
+        subject = f"{cfg['app.title']} - New facility transaction"
+        body = f'{target.text} ({get_app_base_url()}{target.url})'
+
+    elif ressource_type == "favorite_sources":
+        if target.notification_type == "favorite_sources_new_classification":
+            subject = f"{cfg['app.title']} - New classification on a favorite source"
+            body = f'{target.text} ({get_app_base_url()}{target.url})'
+        elif target.notification_type == "favorite_sources_new_spectrum":
+            subject = f"{cfg['app.title']} - New spectrum on a favorite source"
+            body = f'{target.text} ({get_app_base_url()}{target.url})'
+        elif target.notification_type == "favorite_sources_new_comment":
+            subject = f"{cfg['app.title']} - New comment on a favorite source"
+            body = f'{target.text} ({get_app_base_url()}{target.url})'
+
+    elif ressource_type == "mention":
+        subject = f"{cfg['app.title']} - User mentioned you in a comment"
+        body = f'{target.text} ({get_app_base_url()}{target.url})'
+
+    if subject and body and target.user.contact_email:
+        try:
+            send_email(
+                recipients=[target.user.contact_email],
+                subject=subject,
+                body=body,
+            )
+        except Exception as e:
+            log(f"Error sending email notification: {e}")
+
+
+@event.listens_for(UserNotification, 'after_insert')
+def send_sms_notification(mapper, connection, target):
+    if client is None:
+        return
```

Comment:
Can these if blocks be reused between notification types?

Reply:
You're right, some of them can be reused. I'll add a separate function then.

Reply:
Hmmm, I'm actually not so sure if it's worth doing. Most of those if are different actually. Not os many of them can be reused. Depending on the notification type we check different stuff.

Reply:
checking if the user has preferences and such is redundant, so I'll create a function at least for that.

Reply:
Ok that's done.